File: app/src/main/java/com/example/firabasefirstexperience/java/activity/RegisterActivity.java
package com.example.firabasefirstexperience.java.activity;

import android.app.ProgressDialog;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;

import androidx.appcompat.widget.Toolbar;

import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.example.firabasefirstexperience.java.model.User;
import com.example.firabasefirstexperience.R;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

public class RegisterActivity extends AppCompatActivity implements View.OnClickListener {

    private EditText etEmail, etPassword, etUserName;

    private FirebaseAuth auth;
    private DatabaseReference reference;


    private ProgressDialog progressDialog;

    private static final String TAG = "RegisterActivity";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        progressDialog = new ProgressDialog(this);

        etEmail = (EditText) findViewById(R.id.etEmail);
        etPassword = (EditText) findViewById(R.id.etPassword);
        etUserName = (EditText) findViewById(R.id.etUserName);

        Button butSignUp = (Button) findViewById(R.id.butSingUp);

        FirebaseDatabase database = FirebaseDatabase.getInstance();
        reference = database.getReference("Users");
        auth = FirebaseAuth.getInstance();

        Toolbar toolbar = findViewById(R.id.tbForRegisterActivity);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle("Sup Chat");


        butSignUp.setOnClickListener(this);
    }


    public void register(final String eMail, final String password, final String personName, final String imageURL, final String status) {

        progressDialog.setMessage("Registering User...");
        //progressDialog.show();
        auth.createUserWithEmailAndPassword(eMail, password)
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            Toast.makeText(RegisterActivity.this, "User successfully registered!", Toast.LENGTH_SHORT).show();
                            progressDialog.hide();

                            reference.addValueEventListener(new ValueEventListener() {
                                @Override
                                public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                    FirebaseUser firebaseUser = auth.getCurrentUser();

                                    String id = firebaseUser.getUid();

                                    updateUser(id, personName, password, eMail, imageURL, status);

                                }

                                @Override
                                public void onCancelled(@NonNull DatabaseError databaseError) {

                                }
                            });

                            Log.d(TAG, "User successfully registered");

                            finish();
                        } else {

                            Toast.makeText(RegisterActivity.this, "An error occurred , please try later", Toast.LENGTH_LONG).show();
                            Log.d(TAG, "There is an error");
                        }
                    }
                });
    }

    private boolean updateUser(String id, String name, String password, String eMail, String imageURL, String status) {

        reference = FirebaseDatabase.getInstance().getReference("Users").child(id);

        User user = new User(id, name, password, eMail, imageURL, status);
        reference.setValue(user);
        return true;
    }


    @Override
    public void onClick(View v) {

        int id = v.getId();

        switch (id) {
            case R.id.butSingUp:
                final String textUserName = etUserName.getText().toString().trim();
                final String textPassword = etPassword.getText().toString();
                final String textEmail = etEmail.getText().toString().trim();
                final String imageURL = "default";
                final String status = "offline";


                if (TextUtils.isEmpty(textUserName) || TextUtils.isEmpty(textPassword) || TextUtils.isEmpty(textEmail)) {
                    Toast.makeText(RegisterActivity.this, "All fields must be filled", Toast.LENGTH_SHORT).show();
                } else if (textPassword.length() < 8) {
                    Toast.makeText(RegisterActivity.this, "Password must constist at least of 8 characters", Toast.LENGTH_SHORT).show();
                } else {

                    register(textEmail, textPassword, textUserName, imageURL, status);
                }

        }
    }
}

File: app/src/main/java/com/example/firabasefirstexperience/kotlin/RegisterActivity.kt
package com.example.firabasefirstexperience.kotlin

import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.example.firabasefirstexperience.R
import com.example.firabasefirstexperience.java.model.User
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import kotlinx.android.synthetic.main.activity_register_acitivty.*
import kotlinx.android.synthetic.main.bar_layout.*
import kotlinx.android.synthetic.main.bar_layout.view.*
import kotlinx.android.synthetic.main.bar_layout.view.tbForRegisterActivity

class RegisterActivity : AppCompatActivity(), View.OnClickListener {


    private var auth: FirebaseAuth? = null
    private var reference: DatabaseReference? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register_acitivty)


        val database: FirebaseDatabase = FirebaseDatabase.getInstance()
        auth = FirebaseAuth.getInstance()
        reference = database.getReference("Users")

        setSupportActionBar(tbForRegisterActivity)
        supportActionBar?.title = "Sup Chat"
        supportActionBar?.setDisplayHomeAsUpEnabled(true)


        butSingUp.setOnClickListener(this)
    }


    override fun onClick(v: View) {

        val id = v.id

        when (id) {
            R.id.butSingUp -> {
                val textUserName = etUserName.text.toString().trim { it <= ' ' }
                val textPassword = etPassword.text.toString()
                val textEmail = etEmail.text.toString().trim { it <= ' ' }
                val imageURL = "default"
                val status = "offline"


                if (TextUtils.isEmpty(textUserName) || TextUtils.isEmpty(textPassword) || TextUtils.isEmpty(textEmail)) {
                    Toast.makeText(this@RegisterActivity, "All fields must be filled", Toast.LENGTH_SHORT).show()
                } else if (textPassword.length < 8) {
                    Toast.makeText(this@RegisterActivity, "Password must constist at least of 8 characters", Toast.LENGTH_SHORT).show()
                } else {

                    register(textEmail, textPassword, textUserName, imageURL, status)
                }
            }
        }
    }

    private fun register(textEmail: String, textPassword: String, textUserName: String, imageURL: String, status: String) {
        auth?.createUserWithEmailAndPassword(textEmail, textPassword)?.addOnCompleteListener { task ->
            if(task.isSuccessful){
                Toast.makeText(applicationContext, "User Successfully Registered", Toast.LENGTH_SHORT).show()

                reference?.addValueEventListener(object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        val firebaseUser = auth!!.currentUser

                        val id = firebaseUser!!.uid

                        updateUser(id, textUserName, textPassword, textEmail, imageURL, status)

                    }

                    override fun onCancelled(databaseError: DatabaseError) {

                    }
                })

                finish()
        }
            else {
                Toast.makeText(applicationContext, "An error occurred, please try later", Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun updateUser(id: String, textUserName: String, textPassword: String, textEmail: String, imageURL: String, status: String) {
        reference = FirebaseDatabase.getInstance().getReference("Users").child(id)
        val user = User(id, textUserName, textPassword, textEmail, imageURL, status)
        reference?.setValue(user)
    }
}